lab1.py

```diff
@@ -2,51 +2,75 @@
 from redis import Redis, WatchError
 from threading import Thread
 from random import sample
+from collections import Counter
 
-TICKET_NUM_KEY = 'ticket_num'
+UPCOMING_TICKET_ID = 'upcoming_ticket_id'
 
 VALID_NUMBERS = list(range(1, 41))
 
-def draw_ticket_naive(connection: Redis):
-    connection.watch(TICKET_NUM_KEY)
-    ticket_num = int(connection.get(TICKET_NUM_KEY))
-    connection.incr(TICKET_NUM_KEY)
-    connection.unwatch()
-    print(f'Drew ticket #{ticket_num}')
-
-    ticket_numbers = sample(VALID_NUMBERS, 5)
-    ticket_numbers.sort()
-    print(f'Ticket #{ticket_num} numbers: {ticket_numbers}')
-    # We can store the ticket_numbers as key and a list of tickets as value (?)
+def get_winner_tickets(connection: Redis):
+    while True:
+        try:
+            number_of_winner_numbers = int(input('Enter how many numbers to roll: '))
+            break
+        except ValueError:
+            print('Invalid input. Please enter a number.')
 
-    return ticket_num
+    winner_numbers = sample(VALID_NUMBERS, number_of_winner_numbers)
+    winner_numbers.sort()
+    print(f'Winner numbers: {winner_numbers}')
 
-def draw_ticket_slow(connection: Redis):
     with connection.pipeline() as pipe:
-        pipe.watch(TICKET_NUM_KEY)
+        for number in winner_numbers:
+            pipe.watch(number)
         pipe.multi()
-        ticket_num = int(connection.get(TICKET_NUM_KEY))
-        sleep(1)
-        pipe.incr(TICKET_NUM_KEY)
+        for number in winner_numbers:
+            if connection.llen(number) > 0:
+                pipe.lrange(number, 0, -1)
         try:
-            pipe.execute()
+            results = pipe.execute()
+            winner_ticket_ids = [int(ticket_id) for sublist in results for ticket_id in sublist]
+            print(f'Matching numbers of each winning ticket: {dict(Counter(winner_ticket_ids))}')
         except WatchError:
-            print('Another ticket is being updated at the moment. Try drawing again.')
-            return None
+            print('An error occurred while getting winner tickets. Try again.')
 
-    print(f'Drew ticket #{ticket_num}')
+def insert_ticket(ticket_id: int, ticket_numbers: list, connection: Redis):
+    with connection.pipeline() as pipe:
+        for number in ticket_numbers:
+            pipe.watch(number)
+        for number in ticket_numbers:
+            pipe.rpush(number, ticket_id)
```

Comment:
`rpush` is an atomic operation, so there is no need to watch the numbers, the `rpush` will always work even if multiple of them are happening at the same time